smartcontract/contracts/Store.sol

```diff
@@ -6,91 +6,170 @@ import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
 import "@openzeppelin/contracts/utils/Context.sol";
-
 import "./RadaNFTToken.sol";
+import "./RadaToken.sol";
 
 contract Store is Ownable, ReentrancyGuard {
-    RadaNFTToken public radaNFTToken;
-    address public marketplaceToken;
+    
+    RadaNFTToken public nftToken;
+    RadaToken public radaToken;
+
+    struct Campaign {
+        address owner;
+        string name;
+        uint current;
+        uint slot;
+        uint price;
+        uint vote;
+        bool ended;
+    }
+    struct Donor {
+        address add;
+        uint amount;
+        bool voted;
+    }
+    struct Distributer {
+        address add;
+        uint amount;
+    }
 
-    mapping(uint256 => uint256) public itemPrices;
+    
+    mapping(address => bool) public ableToCreateCampaign;
+    mapping(uint => Donor[]) public donorsOfCampaign;
+    mapping(uint => Distributer[]) public distributersOfCampaign;
+    mapping(uint => Campaign) public campaigns;
+    
+    uint public campaign_count;
 
-    event Listing(address indexed owner, uint256 indexed nftId, uint256 price);
 
-    event Unlisting(address indexed owner, uint256 indexed nftId);
 
-    event Purchase(
-        address indexed previousOwner,
-        address indexed newOwner,
-        uint256 indexed nftId,
-        uint256 listingPrice
-    );
+    modifier isOwnerOfCampaign(address _add, uint _campaign_id) {
+        require(_add == campaigns[_campaign_id].owner);
+        _;
+    }
+
+    modifier isWorkingCampaign(uint _campaign_id){
+        require(campaigns[_campaign_id].slot > 0);
+        _;
+    
+    }
+    modifier isEndedCampaign(uint _campaign_id){
+        require(campaigns[_campaign_id].ended);
+        _;
+    }
+    modifier isAcceptedCampaign(uint _campaign_id) {
+        require(campaigns[_campaign_id].vote > 0);
+        _;
+    }
+    modifier isDonorOfCampaign(address _add, uint _donor_id, uint _campaign_id){
+        require(
+            _add ==donorsOfCampaign[_campaign_id][_donor_id].add, 
+            "Donor id not matching donor address"
+        );
+        _;
+    }
+    modifier isDonorVotedForCampaign(uint _donor_id, uint _campaign_id){
```

Comment:
Donor can vote multiple times

Reply:
Donor can vote only one time per a distributer list because vote one-donor for one-distributer is complicated